components/client/src/main/java/com/hotels/styx/client/applications/metrics/OriginMetrics.java

```diff
@@ -40,114 +41,100 @@
  * connections scheduled on different event loops.
  */
 public class OriginMetrics implements OriginStats {
-    private static final Logger LOG = LoggerFactory.getLogger(OriginMetrics.class);
-    private static final int SERVER_ERROR_CLASS = 5;
+    public static final String ORIGIN_TAG = "origin";
+    public static final String APP_TAG = "appId";
+    public static final String STATUS_TAG = "statusCode";
+    public static final String STATUS_CLASS_TAG = "statusClass";
 
-    private final ApplicationMetrics applicationMetrics;
+    public static final String SUCCESS_COUNTER_NAME = "request.success";
+    public static final String FAILURE_COUNTER_NAME = "request.error";
+    public static final String STATUS_COUNTER_NAME = "response.status";
+    public static final String CANCELLATION_COUNTER_NAME = "request.cancellation";
+    public static final String LATENCY_TIMER_NAME = "request.latency";
+    public static final String TTFB_TIMER_NAME = "request.timetofirstbyte";
 
-    private final String requestMetricPrefix;
+    private static final Duration MIN_HISTOGRAM_BUCKET = of(10, MILLIS);
+    private static final Duration MAX_HISTOGRAM_BUCKET = of(5, MINUTES);
```

Comment:
The histogram bucket sizing must be configurable. Seriously.

For example MIN_HISTOGRAM_BUCKET way too high. In AWS load test environment 99.9% of the requests would complete within 10ms. The 75th percentile is around 2ms. These figures are with a test redirect rule engine plugin. Removing this plugin would only speed things up. Sadly I don't have any data with bare styx without plugins.

Depending on the computing environment, instance sizing, origins, and plugin configuration, we could even see sub-millisecond latencies.

Also the max value can be influenced by many factors:
* Database queries in the backend service feeding more HTTP content, over the chunked transfer encoding, as the query progresses
* External payment processors, etc.
* Streaming file downloads/uploads.
* End user "last mile" link speeds. For example a slow mobile connection could trigger Styx flow control on the origin side and so inflate latencies.

Reply:
Choosing the boundaries for the upper and lower buckets is a trade-off between capturing all possible values and controlling the number of time series created.

Should the percentile being calculated fall outside of the range then it will be capped to the range. We should note that in the documentation.

Good point about the lower bound. The value should be lowered and I'll amend it to 1ms. At that point if the percentile you're looking for has a value outside the range then it will be clipped to the 1ms - 5min range. You would still have the mean and max values.

If, once we try out these values, there is a need to change the value per application, origin or host-proxy then we should raise a ticket to extend the configuration. It might also make sense in the near future to allow the user to specify SLOs so that they can measure their compliance. But I think that belongs in a follow up PR.